src/Elastic.Apm/OpenTelemetry/ElasticActivityListener.cs

```diff
@@ -0,0 +1,173 @@
+						distributedTracingData: dt);
+				}
+				else if (activity.ParentId == null)
+				{
+					transaction = _tracer.StartTransactionInternal(activity.DisplayName, "unknown",
+						TimeUtils.ToTimestamp(activity.StartTimeUtc), true, activity.SpanId.ToString(),
+						activity.TraceId.ToString());
+				}
+				else
+				{
+					Span newSpan;
+					if (_tracer.CurrentSpan == null)
+					{
+						newSpan = (_tracer.CurrentTransaction as Transaction)?.StartSpanInternal(activity.DisplayName, "unknown",
+							timestamp: TimeUtils.ToTimestamp(activity.StartTimeUtc), id: activity.SpanId.ToString());
+					}
+					else
+					{
+						newSpan = (_tracer.CurrentSpan as Span)?.StartSpanInternal(activity.DisplayName, "unknown",
+							timestamp: TimeUtils.ToTimestamp(activity.StartTimeUtc), id: activity.SpanId.ToString());
+					}
+
+					if (newSpan != null)
+					{
+						newSpan.Otel = new OTel();
+						newSpan.Otel.SpanKind = activity.Kind.ToString();
+
+						if (activity.Id != null) ActiveSpans.TryAdd(activity.Id, newSpan);
+					}
+				}
+
+				if (transaction != null)
+				{
+					transaction.Otel = new OTel();
+					transaction.Otel.SpanKind = activity.Kind.ToString();
+
+					if (activity.Kind == ActivityKind.Server)
+						transaction.Type = ApiConstants.TypeRequest;
+					else if (activity.Kind == ActivityKind.Consumer) transaction.Type = ApiConstants.TypeMessaging;
+
+					if (activity.Id != null) ActiveTransactions.TryAdd(activity.Id, transaction);
+				}
+			};
+
+
+		private Action<Activity> ActivityStopped =>
+			activity =>
+			{
+				if (activity == null)
+				{
+					_logger.Trace()?.Log("ActivityStopped called with `null` activity. Ignoring `null` activity.");
+					return;
+				}
+
+				_logger.Trace()?.Log($"ActivityStopped: name:{activity.DisplayName} id:{activity.Id} traceId:{activity.TraceId}");
+
+				if (KnownListeners.KnownListenersList.Contains(activity.DisplayName))
+					return;
+
+				if (activity.Id != null && ActiveTransactions.TryRemove(activity.Id, out var transaction))
+				{
+					transaction.Duration = activity.Duration.TotalMilliseconds;
+
+					if (activity.Tags.Any())
+						transaction.Otel.Attributes = new Dictionary<string, string>();
+
+					foreach (var tag in activity.Tags) transaction.Otel.Attributes.Add(tag.Key, tag.Value);
+					transaction.End();
+				}
+
+				if (activity.Id != null && ActiveSpans.TryRemove(activity.Id, out var span))
+				{
+					span.Duration = activity.Duration.TotalMilliseconds;
+
+					if (activity.Tags.Any())
+						span.Otel.Attributes = new Dictionary<string, string>();
+
+					foreach (var tag in activity.Tags) span.Otel.Attributes.Add(tag.Key, tag.Value);
+
+					InferSpanTypeAndSubType(span, activity);
+					span.End();
+				}
+			};
+
+		public ActivityListener listener { get; }
+
+		private void InferSpanTypeAndSubType(Span span, Activity activity)
+		{
+			if (activity.Kind == ActivityKind.Client && activity.Tags.Any(n => n.Key == "http.url" || n.Key == "http.scheme"))
```

Comment:
I think this could filter on activity.Kind first, then check tags, to avoid checking kind on each condition

Reply:
I changed it as you suggested, after looking at it, this (at least to me) hurts readability. Ideally this would be part of the spec and I think there would be more advantage to write it the same way as the spec - e.g. for maintenance. I doubt this makes any impact on perf. 

For now let's keep it the way you suggest - I may revisit this with benchmarks if the spec evolves and this part becomes hard to read.